### ISSUE TITLE
Could someone help me to clarify the concept of multi-node training for webdataset ?

### ISSUE THREAD

**Issue by HuangChiEn** [webdataset/webdataset#340]
i have saw the [reply](https://github.com/webdataset/webdataset/issues/250#issuecomment-1475780909) mentioned some concept of multi-node training.

However, I can not verify my understanding is correct or not for the following concepts : 
> Let's say we have 2 gpus with DDP mode
1. If `resample=True`, each gpu will **randomly** fetch the batch data from the webds. Since fetching behavior is random, it **may** lead to different devices getting the same data (at least, not always getting the same data). So, if I run 1 epoch in a 2 gpus setting, it should be equivalent to running **2 epoch** for single gpu settings ?
2. `with_epoch(U)` means the epoch length. That's `U = num_of_data // batch_size` (in single gpu case). However, I wonder that the exactly U is hard to get it. We use webdataset due to its' large scale, then why should we back to calculate the dataset size. I hope the setup for U can be approximated and not affect the training settings to much.
3. `with_length` sets the nominal length of the dataset (the value returned by len(...)), but doesn't change anything else about the pipeline. So, can i ignore it, or i still need to give it a dummy value ?
4. `repeat(n)` repeats the iterator n times (i.e., it runs for n epochs). But the code is still running without this settings. What is the default behavior for this argument ?

Any suggestion will be appreciated !!

**Comment by tmbdev** [comment]
You have three options:

(1) If you want multinode training AND you want exactly one permutation of the dataset per epochs, you should use `wids` (distributed with WebDataset). Here is an example:

https://github.com/webdataset/webdataset/blob/main/examples/train-resnet50-multiray-wids.ipynb

Using anything based on `IterableDataset` for this use case is quite difficult because of the way PyTorch's DataLoader functions.

(2) If you want the most efficient multinode training and don't care about having exactly one permutation of the dataset per epoch, you should use `WebDataset` with resampling. This also greatly simplifies restarting.

(3) If you have more complex preprocessing/distributed computing needs, you have the option of using `ray.data` for data preparation, which allows you to do full sorting/shuffling/distributed computing as part of your input pipeline. Ray will also handle distributed training itself for you.

**Comment by HuangChiEn** [comment]
> You have three options:
> 
> (1) If you want multinode training AND you want exactly one permutation of the dataset per epochs, you should use `wids` (distributed with WebDataset). Here is an example:
> 
> https://github.com/webdataset/webdataset/blob/main/examples/train-resnet50-multiray-wids.ipynb
> 
> Using anything based on `IterableDataset` for this use case is quite difficult because of the way PyTorch's DataLoader functions.
> 
> (2) If you want the most efficient multinode training and don't care about having exactly one permutation of the dataset per epoch, you should use `WebDataset` with resampling. This also greatly simplifies restarting.
> 
> (3) If you have more complex preprocessing/distributed computing needs, you have the option of using `ray.data` for data preparation, which allows you to do full sorting/shuffling/distributed computing as part of your input pipeline. Ray will also handle distributed training itself for you.

Yeah, but the problem is we can not understand the **behavior** of the multi-gpu distributed training with webdataset. Let's say we apply `resample=True` and `with_epoch(.)`.





> You have three options:
> 
> (1) If you want multinode training AND you want exactly one permutation of the dataset per epochs, you should use `wids` (distributed with WebDataset). Here is an example:
> 
> https://github.com/webdataset/webdataset/blob/main/examples/train-resnet50-multiray-wids.ipynb
> 
> Using anything based on `IterableDataset` for this use case is quite difficult because of the way PyTorch's DataLoader functions.
> 
> (2) If you want the most efficient multinode training and don't care about having exactly one permutation of the dataset per epoch, you should use `WebDataset` with resampling. This also greatly simplifies restarting.
> 
> (3) If you have more complex preprocessing/distributed computing needs, you have the option of using `ray.data` for data preparation, which allows you to do full sorting/shuffling/distributed computing as part of your input pipeline. Ray will also handle distributed training itself for you.

Thanks for your reply, we want to go option (2), but we have no idea how to properly setup n for `with_epoch(n)`.

Let's say we have one node 2 gpus, batch_size=32, (then global_batch_size = 64), num_worker=16 (torch or web loaders' worker for speedup data-loading). 
The dataset has 256 samples, and those samples are wrapped in 128 shards tar-file (each shard has 2 samples) 

**If I want to run 2 step per epoch**, should I setup like this
`wds.Webdataset(..., resample=True).(..).with_epoch(batch_size * 2)` ?

Could you help me to verify if my understanding is correct for some **miscellnous concepts** :
1. In above example, each worker will load 2 shard tar-file (4 samples) to contribute 64 data, since resample=True, each worker will load it **randomly**  ?

2. suppose we setup `n=batch_size*n_step`, finally it will run n_step by calculating `n // batch_size` ?
